Select script type while receiving

turbolay opened the pull request:
This PR lets the user decide which ScriptType to use when creating a receiving address, and it displays this ScriptType in the `Addresses awaiting payment` view.

In this view, I was forced to shorten Taproot addresses to the length of segwit ones because taproot addresses are too big and with the fixed size of the views in Wasabi it was not fitting and introduced an ugly scrollbar

This PR contains several known problems:
- Displaying the ScriptType in the ReceiveView (the one with the QRCode) doesn't work for some reason
- Taproot address is cut in the display window
- It's still ugly in Address Awaiting Payment screen when screen is as small as possible

Missing:
Store last choice

turbolay commented on WalletWasabi.Fluent/Views/Wallets/Receive/ReceiveView.axaml:
```diff
@@ -43,6 +45,18 @@
           </Interaction.Behaviors>
         </TagsBox>
+        <DockPanel Margin="0,20,0,0" HorizontalAlignment="Left">
+          <InfoMessage DockPanel.Dock="Right" VerticalAlignment="Center" ToolTip.Tip="Taproot offers generally better fees and privacy but SegWit has wider compatibility." />
```
@Kruwed @MarnixCroes Any idea?

MarnixCroes replied:
```suggestion
          <InfoMessage DockPanel.Dock="Right" VerticalAlignment="Center" ToolTip.Tip="Taproot generally offers lower fees and better privacy but SegWit has wider compatibility." />
```
yeah it's ok.

yahiheb reviewed (changes requested):
- I would make the script type combo box slightly longer so it won't resize each time a different script type is chosen.
- > In this view, I was forced to shorten Taproot addresses to the length of segwit ones because taproot addresses are too big and with the fixed size of the views in Wasabi it was not fitting and introduced an ugly scrollbar

  Copying the shortened Taproot address copies the displayed text, with the dots.

turbolay commented:
Getting closer, I'm now using a SubAction. 
I initially tried to do it one extra step to avoid separating Segwit and Taproot like this but I didn't manage to make the NextButton a SubActionButton

Known issues:
- Need a tooltip to display full addresses for taproot without copying
- My dynamic icon doesn't work and is always Segwit

yahiheb reviewed (commented):
The suggestions mentioned here https://github.com/WalletWasabi/WalletWasabi/pull/13294#pullrequestreview-2211011515 are now fixed.
Tested this and it LGTM, except the known issues mentioned by Clement.

Note that now the segwit and taproot addresses are displayed separately on the receive dialog depending on which script type is chosen.

Kruwed commented:
Will this break hardware wallets? Or is it a hot wallet exclusive feature like RBF is?

turbolay commented:
> Will this break hardware wallets? Or is it a hot wallet exclusive feature like RBF is?

Good point, I have to check from the `KeyManager` which script types are available

turbolay commented:
What @Kruwed mentioned is now fixed, when Taproot is not available on the wallet it now displays a simple button, as before this PR, except that the icon will be `[SW]` for consistency.

I also fixed my dynamic icon by writing the most ridiculous code ever seen by mankind, but at least it made the code really easy and located instead of having to change lot of code everywhere and spray my bad code.

I sincerely apologize to all Avalonia devs that will be hurt looking at my awful code.

Everything is done except the tooltip, I will add the feature, clean as much as I can code for this PR and merge.

turbolay commented:
Tooltip added. It does not work for some reason in the TreeDataGrid so I only added it in the `ReceiveAddressView` (the one with the QR code). I think it's enough

I will try to improve the code and then merge

@Kruwed @yahiheb @MarnixCroes UX/Behavior is final, if you want to test and provide feedback, feel free.

turbolay commented on WalletWasabi.Fluent/ViewModels/Wallets/WalletViewModel.cs:
```diff
@@ -237,7 +265,7 @@ private ISearchItem[] CreateSearchItems()
 		return new ISearchItem[]
 		{
-			new ActionableItem("Receive", "Display wallet receive dialog", () => { ReceiveCommand.ExecuteIfCan(); return Task.CompletedTask; }, "Wallet", new[] { "Wallet", "Receive", "Action", }) { Icon = "wallet_action_receive", IsDefault = true, Priority = 2 },
+			new ActionableItem("Receive", "Display wallet receive dialog", () => { DefaultReceiveCommand.ExecuteIfCan(); return Task.CompletedTask; }, "Wallet", new[] { "Wallet", "Receive", "Action", }) { Icon = "wallet_action_receive", IsDefault = true, Priority = 2 },
```
Searching for "Receive" in the search bar will always initiate the flow with the last script type selected.

yahiheb reviewed (approved):
tACK

MarnixCroes reviewed (commented):
thanks for the effort to use the subaction button, it is much better.

the sorting at `Addresses Awaiting Payment` dialog doesn't work anymore, both for _label_ and _address_.

MarnixCroes commented:
when releasing with this feature we should also have it available in the RPC https://github.com/WalletWasabi/WalletWasabi/issues/12176

lontivero reviewed (dismissed):
tACK

MarnixCroes reviewed (approved):
tack 04c711a7cfa4eb139f27514ef348473d7c598bfa

tested GUI & RPC

---

only one note/question is about the _Default_, this is different from the send button. Where the default always stays the same. For receive the default is last used.
This difference in behavior may be odd.

I think it is ok to remove this behavior and to just default to segwit? (can be changed to taproot by default later)

yahiheb reviewed (approved):
tACK

Kruwed commented:
I don't feel that displaying [SW] and [TR] on the Receive button is necessary since the green dot already tracks which is selected as the default. The boxed in text isn't very aesthetic, and it increases the cognitive load on users who have to squint to read it compared to the previous receive icon (which also nicely matches the transaction history icon).

turbolay commented:
Given that my extra UX with animated icon and storage of last choice didn't make unanimity, I reverted it to a simpler UX and code. If you feel nostalgic about my beautiful animated button and previous UX feel free to merge master on this branch, revert 7a963df and open a PR
